source/opt/ir_context.h

```diff
@@ -121,8 +133,61 @@ class IRContext {
   // Appends a function to this module.
   inline void AddFunction(std::unique_ptr<Function>&& f);
 
+  // Returns a pointer to a def-use manager.  If the def-use manager is
+  // invalid, it is rebuilt first.
+  opt::analysis::DefUseManager* get_def_use_mgr() {
+    if (!AreAnalysesValid(kAnalysisDefUse)) BuildDefUseManager();
+    return def_use_mgr_.get();
+  }
+
+  // Builds the def-use manager from scratch, even if it was already valid.
+  void BuildDefUseManager() {
+    def_use_mgr_.reset(new opt::analysis::DefUseManager(module()));
+    valid_analyses_ = (Analysis)(valid_analyses_ | kAnalysisDefUse);
+  }
+
+  // Sets the message consumer to the given |consumer|. |consumer| which will be
+  // invoked every time there is a message to be communicated to the outside.
+  void SetMessageConsumer(spvtools::MessageConsumer c) {
+    consumer_ = std::move(c);
+  }
+
+  // Returns the reference to the message consumer for this pass.
+  const spvtools::MessageConsumer& consumer() const { return consumer_; }
+
+  // Rebuilds the analyses in |set| that are invalid.
+  void BuildAnalyses(Analysis set);
+
+  // Invalidates all of the analyses except for those in |preserved_analyses|.
+  void InvalidateAnalyses(Analysis preserved_analyses);
+
+  // Turns the instruction defining the given |id| into a Nop. Returns true on
```

Comment:
For the long term:  Since we now have easy deletion of instructions, should we change this to mean deletion of the instruction?  Changing it to no-op was an artefact of not wanting to manipulate the vector of instructions.

Reply:
Yes, I looked into doing that.  The problem is that current code is written assuming the current instruction iterator is not invalidated after calling killing the instruction.  To fix that, we need to rewrite that code.  I did not want to muddy the waters in this CL by doing that as well.

Reply:
+1